BCF entry point with intensity and contamination checks using BCF for data_catalog usage.

rajwanir opened the pull request:
1. Creates BCF entry point for data catalog.
2.  Makes BPM manifest optional (relevant for aggregated file inputs e.g. BCF, VCF or BED)
3. Adds scripts to compute median intensity and contamination score from BCF.
4. Separates the intensity checks from contamination checks.
5. No observable difference with standard GTC input.

Commit:
makes illumina_manifest_file optional in the config

Commit:
Uses snps_array name instead of bpm to name *.abf.txt
Avoids dependency of bpm to name allele B frequencies (abf) file.

Commit:
Adds a new idat_intensity workflow

to separate median idat intensity retrieval from verifyIDintensity bundled into contamination.smk

Commit:
Adds vcf2abf.py and vcf2adpc.py as prep scripts for contamination checks with vcf entry

Commit:
Creates a BCF entry point.
Modifis the entry_points.smk to create BCF entry point by simply converting BCF to plink BED. Testing and validation yet to be done.

Commit:
Removes IDAT intensity checks from contamination.smk
A previous commit puts them in a separate idat_intensity.smk

Commit:
Adds rules contamination.smk to do contamination checks with bcf input.

Commit:
Enables contamiantion check with bcf input in cluster mode.
Modifies contamination.smk and grouped_contamination.py to enable contamination check in cluster mode.

Commit:
Adds capability to infer median intensity directly from input BCF/VCF.
Avoids the dependency on IDAT files for calculating median intensity with VCF/BCF input.
Adds scripts for both in per-sample and grouped/cluster mode.
Modifies the intensity workflow to execute appropriately if VCF/BCF input.

Commit:
Adds the distinction of contamination and intensity checks in the snakefile and sample_qc subworkflow
Other than existing 'use_contamination' checks, also adds 'intensity_retreived' and 'contamination_checked'
tests which simply tests specifically if output csv files were created regardless of configs/entry point to feed
them to sample_qc.

Commit:
Correcting conditions such with BCF input intensity and contamination check is default

jaamarks commented on src/cgr_gwas_qc/workflow/sub_workflows/entry_points.smk:
```diff
@@ -227,3 +223,40 @@ elif cfg.config.user_files.bed and cfg.config.user_files.bim and cfg.config.user
+        file for processing.
+        """
+        input:
+            vcf=cfg.config.user_files.bcf,
```
Should the input variable name be `bcf`?

rajwanir replied:
Renamed in 2d7964037ccb890f4fd39c7093444455d8126646 and 5009ca51c62d23cb865d0596f697f8456999762e

jaamarks commented on src/cgr_gwas_qc/workflow/sub_workflows/entry_points.smk:
```diff
@@ -227,3 +223,40 @@ elif cfg.config.user_files.bed and cfg.config.user_files.bim and cfg.config.user
+        shell:
+            "plink "
+            "--allow-extra-chr 0 --keep-allele-order --double-id "
+            "--bcf {input.vcf} "
```
`{input.bcf}`

rajwanir replied:
Renamed in 2d7964037ccb890f4fd39c7093444455d8126646 and 5009ca51c62d23cb865d0596f697f8456999762e

jaamarks reviewed (commented):
I see that this removes the IDAT intensity checks from contamination.smk and creates a separate subworkflow. While I understand that modularity can be beneficial, my first inclination is that we wouldn't want to separate this into distinct subworkflows. 
It seems that keeping one unifying contamination workflow is more inline with simplicity.

Could you please elaborate on the benefits of the current approach? I’m interested in understanding how this separation improves the workflow and if there are specific advantages that we should consider.

jaamarks commented on src/cgr_gwas_qc/workflow/modules/thousand_genomes.smk:
```diff
@@ -28,6 +28,31 @@ rule pull_b_allele_freq_from_1kg:
+    ``AF`` which ignores super population.
+    """
+    input:
+        vcf_file=cfg.config.user_files.bcf,
```
should this variable name be `bcf_file`?

rajwanir replied:
Renamed in 2d7964037ccb890f4fd39c7093444455d8126646 and 5009ca51c62d23cb865d0596f697f8456999762e

rajwanir replied:
Hi @jaamarks ,

Thanks again for taking out the time to review this PR. Appreciate your review in detail and checking for any issues. Overall, I think updated branch should reflect all/most changes requested. If anything else needed to be fixed, please feel free to let me know. It would be great if we have this merged into the next release too if possible. Might save sometime in realigning and resolving conflicts with new release. However, if you see any concerns in merging, we can hold on and do it later too.

Here are summary of the changes:

1. Rebased the branch as suggested.
2. Renamed variable to bcf_file wherever possible. Actually, the idea is that all functionality should be supported with vcf or bcf input seemlessly. And for the most part it should. However, I haven't done extensive testing with vcf_file. And atleast the plink import requires an explict --vcf or --bcf switch. For now, I have followed your suggestion to rename to bcf_file for clarity. Later on, we can test if vcf for all the worklows and if it works then we can revert the variable name to vcf_file for simplicity.
3. As discussed in the meeting, separating intensity subworkflow from contamination appeared more logical to me since the intensity checks specifically checks for intensity and not contamination. There is not much common between the two. Separating them also allows for testing and optimizing them independently if needed.

Feel free to let me know anything else needed. Thank you very much.

jaamarks commented on src/cgr_gwas_qc/workflow/scripts/vcf2adpc.py:
```diff
@@ -27,7 +27,7 @@ def main(
     vcf_file: Path = typer.Argument(
         ..., help="Path to a multisample VCF file with needed scores.", exists=True, readable=True
     ),
-    target_sample: str = typer.Argument(..., help="Sample name.", exists=True, readable=True),
```
Why are we removing these defaults?

rajwanir replied:
The `target_sample` here is a string here not a file path checking for for existence or readable is not relevant. Essentially, the functionality of this `vcf2adpc `is such that it pulls the information for the `target_sample `from the aggregated `bcf ` input file and writes to `adpc.bin` (outfile) for contamination workflow.

jaamarks commented on src/cgr_gwas_qc/workflow/sub_workflows/contamination.smk:
```diff
@@ -28,7 +28,16 @@ module thousand_genomes:
+
+    use rule pull_b_allele_freq_from_1kg_bcfinput from thousand_genomes as pull_b_allele_freq_from_1kg with:
+        input:
+            vcf_file=cfg.config.user_files.bcf,
```
Should this variable name be `bcf_file`?

rajwanir replied:
Renamed in 5009ca51c62d23cb865d0596f697f8456999762e

jaamarks reviewed (changes requested):
Hi @rajwanir,

Thank you for the PR! I’ve completed my review of your commits, and now I need to conduct some testing to ensure everything works as expected. I have a few questions and comments that I’d like to summarize here:

1. Could you please fetch the latest changes from the default branch and rebase (instead of merging)? This approach helps maintain a cleaner history.
2. I noticed a few instances where the variable name vcf_file was used, and I wondered if it should be bcf_file instead. My main concern is whether this might cause confusion for developers, or if it’s common practice to refer to BCF files as VCF since BCF is essentially the binary version of VCF. What are your thoughts on this?
3. I have some concerns about splitting the contamination subworkflow. Could you address this comment and discuss the pros and cons of this approach?

Thank you!

jaamarks commented:
Hi @rajwanir,

Thank you for the PR! I’ve completed my review of your commits, and now I need to conduct some testing to ensure everything works as expected. I have a few questions and comments that I’d like to summarize here:

1. Could you please fetch the latest changes from the default branch and rebase (instead of merging)? This approach helps maintain a cleaner history.
2. I noticed a few instances where the variable name vcf_file was used, and I wondered if it should be bcf_file instead. My main concern is whether this might cause confusion for developers, or if it’s common practice to refer to BCF files as VCF since BCF is essentially the binary version of VCF. What are your thoughts on this?
3. I have some concerns about splitting the contamination subworkflow. Could you address this comment and discuss the pros and cons of this approach?

Thank you!

jaamarks commented on src/cgr_gwas_qc/models/config/user_files.py:
```diff
@@ -89,6 +89,12 @@ class UserFiles(BaseModel):
+        None,
+        description="The full path to an aggregated BCF/VCF file perferably encoding the GenCall scores.",
+    )
+
```
Please add a validator for BCF. Use the `validate_gtc_pattern` as a template.

rajwanir replied:
Added validator in 52f97c790a3d3fb43ec28167757e21cf95ca71a2 to ensure the input `bcf `ends in `.bcf`

Commit:
Renames variable vcf_file to bcf_file when bcf file is refered.

jaamarks commented on src/cgr_gwas_qc/workflow/sub_workflows/contamination.smk:
```diff
@@ -67,195 +65,249 @@ rule verifyidintensity_conda:
+    else:
+
+        rule per_sample_vcf_to_adpc:
+            """Converts a sample's GTC/BPM to an Illumina ADPC.BIN.
```
This docstring needs to be updated to reflect that the input is a BCF not GTC/BPM.

rajwanir replied:
Updated in f6ed2d7bf1f5213b6a1f5b793c3cd0c4321f0e5d

jaamarks commented:
I am encountering the following error with the `pull_b_allele_freq_from_1kg` rule. Could you please help address this error message?

```
[Wed Sep 11 14:21:26 2024]
rule pull_b_allele_freq_from_1kg:
    input: /mnt/nfs/gigantor/ifs/DCEG/Home/marksja/dev-tests/samples.bcf, /DCEG/Projects/DataDelivery/GWAS_Data_Delivery/CONFLUENCE_FB28648_05172021/testing_500samples/manifest/ALL.wgs.shapeit2_integrated_v1a.GRCh38.20181129.sites.vcf.gz
    output: sample_level/GSAv3Confluence.AF.abf.txt
    jobid: 0
    resources: mem_mb=19787, disk_mb=19787, tmpdir=/tmp                                                                       
/home/marksja/miniconda3/envs/cgr-dev/bin/python /mnt/nfs/gigantor/ifs/DCEG/Home/marksja/dev-tests/024/.snakemake/scripts/tmprlcuev93.vcf2abf.py
[E::idx_find_and_load] Could not retrieve index file for '/mnt/nfs/gigantor/ifs/DCEG/Home/marksja/dev-tests/samples.bcf'
Traceback (most recent call last):
  File "/mnt/nfs/gigantor/ifs/DCEG/Home/marksja/dev-tests/024/.snakemake/scripts/tmprlcuev93.vcf2abf.py", line 144, in <module>
    main(**defaults)  # type: ignore
  File "/mnt/nfs/gigantor/ifs/DCEG/Home/marksja/dev-tests/024/.snakemake/scripts/tmprlcuev93.vcf2abf.py", line 77, in main
    rec.info["GC_SCORE"],
  File "pysam/libcbcf.pyx", line 2598, in pysam.libcbcf.VariantRecordInfo.__getitem__
KeyError: 'Unknown INFO field: GC_SCORE'
[Wed Sep 11 14:21:27 2024]                                                                                                    Error in rule pull_b_allele_freq_from_1kg:
    jobid: 0                                                                                                                      output: sample_level/GSAv3Confluence.AF.abf.txt
```

Thanks!

jaamarks commented:
There are two identical files in the sub_workflows directory now.
* idat_intensity.smk
* intensity_check.smk
```
sub_workflows$ ls
contamination.smk  delivery.smk  entry_points.smk  idat_intensity.smk  intensity_check.smk  sample_qc.smk  subject_qc.smk

```

jaamarks commented:
> No observable difference with standard GTC input.

I don't observe any difference in the QC_Report.docx. But we do see differences in the corresponding Plink files created (BED/BIM/FAM). For example:
```
$ diff samples.bim ../../020a/sample_level/samples.bim | head
5c5
< 1     rs3131972       0       817341  G       A
---
> 1     rs3131972       0       817341  A       G

```

We see the alleles flipped. Can you address this?

Commit:
Removes the duplicated subworkflow file for intensity checks.
Removes idat_intensity.smk and keeps intensity_check.smk

Commit:
Updates the docstring in contamination.smk to indicate when BCF file is used.

Commit:
Renames input variable to bcf_file.
Renames from vcf_file to bcf_file explictly indicate that bcf is input.

Commit:
Imports snakemake params all at once in grouped_contamination.py
snakemake params were imported through named import. Changed to import all params through a loop in unnamed fashion. Allows seemless compatibility when gtc or bcf entry point is used.

Commit:
Removes typo duplicated header in entry_points
The starting few lines were duplicated in the entry_points in copy/paste. This removes the duplicated lines.

rajwanir commented:
> > No observable difference with standard GTC input.
> 
> I don't observe any difference in the QC_Report.docx. But we do see differences in the corresponding Plink files created (BED/BIM/FAM). For example:
> 
> ```
> $ diff samples.bim ../../020a/sample_level/samples.bim | head
> 5c5
> < 1     rs3131972       0       817341  G       A
> ---
> > 1     rs3131972       0       817341  A       G
> ```
> 
> We see the alleles flipped. Can you address this?

This seems a flip on the gtc-to-vcf workflow when it imports to plink rather than on bcf entry point with bcf to plink. In the bcf entry point I have protected the allele order with `--keep-allele-order` switch so allele order would always be preserved from bcf to plink conversion. The bcf/vcf encodes alleles in the REF/ALT space but plink bim file encodes in the major/minor space. If we want to preserve the VCF allele order and go with the typical notion that REF allele is major allele. So,

vcf/bcf record (CHROM ID REF ALT):
`chr1 rs3131972 A G
`
should translated into bim record (CHROM ID CM POS ALLELE1(minor) ALLELE2 (major)):
`1       rs3131972       0       817341  G       A`

bim format is described here: https://www.cog-genomics.org/plink/1.9/formats#bim

Another simpler and emperical way to verify this is if you re-export samples_level/samples.bim to vcf, you would see the orignal allele order as in vcf. 

You can reexport plink bed/bim/fam to vcf.gz with the following:
`plink --bfile bcf_out/samples --recode vcf-iid bgz --out bcf_out/samples --allow-extra-chr --keep-allele-order --const-fid 0
`
And compare the plink rexported with the orignal input like this:
`diff <(bcftools query -f"%ID %REF %ALT" /scratch/marksja/samples.bcf |sort) <(bcftools query -f"%ID %REF %ALT" bcf_out/samples.vcf.gz |sort)`

You should expect to see no difference except for multi-allellics not imported which is not supported by plink/1.9 and would need to be fixed upstream in preparation of bcf.

Overall, I can open a separate issue and we can adress this more holistically throughout the pipeline in a different PR. However, for the BCF entry point, I  see the transformation is as expected with no flip.

Feel free to let me know if this seems acceptable.

Commit:
Shifts to using IGC score instead of gentrain score in vcf2adpc.py for contamination checks.
Previously GC_SCORE was added to the adpc.bin which had depenency that a cluster egt file had to be used in preparation of vcf/bcf. IGC score is encoded in gtc so doesn't depended on cluster egt file. The conamination scores should also be more similar with the gtc input.

rajwanir commented:
> I am encountering the following error with the `pull_b_allele_freq_from_1kg` rule. Could you please help address this error message?
> 
> ```
> [Wed Sep 11 14:21:26 2024]
> rule pull_b_allele_freq_from_1kg:
>     input: /mnt/nfs/gigantor/ifs/DCEG/Home/marksja/dev-tests/samples.bcf, /DCEG/Projects/DataDelivery/GWAS_Data_Delivery/CONFLUENCE_FB28648_05172021/testing_500samples/manifest/ALL.wgs.shapeit2_integrated_v1a.GRCh38.20181129.sites.vcf.gz
>     output: sample_level/GSAv3Confluence.AF.abf.txt
>     jobid: 0
>     resources: mem_mb=19787, disk_mb=19787, tmpdir=/tmp                                                                       
> /home/marksja/miniconda3/envs/cgr-dev/bin/python /mnt/nfs/gigantor/ifs/DCEG/Home/marksja/dev-tests/024/.snakemake/scripts/tmprlcuev93.vcf2abf.py
> [E::idx_find_and_load] Could not retrieve index file for '/mnt/nfs/gigantor/ifs/DCEG/Home/marksja/dev-tests/samples.bcf'
> Traceback (most recent call last):
>   File "/mnt/nfs/gigantor/ifs/DCEG/Home/marksja/dev-tests/024/.snakemake/scripts/tmprlcuev93.vcf2abf.py", line 144, in <module>
>     main(**defaults)  # type: ignore
>   File "/mnt/nfs/gigantor/ifs/DCEG/Home/marksja/dev-tests/024/.snakemake/scripts/tmprlcuev93.vcf2abf.py", line 77, in main
>     rec.info["GC_SCORE"],
>   File "pysam/libcbcf.pyx", line 2598, in pysam.libcbcf.VariantRecordInfo.__getitem__
> KeyError: 'Unknown INFO field: GC_SCORE'
> [Wed Sep 11 14:21:27 2024]                                                                                                    Error in rule pull_b_allele_freq_from_1kg:
>     jobid: 0                                                                                                                      output: sample_level/GSAv3Confluence.AF.abf.txt
> ```
> 
> Thanks!

270e1055378ac20ada4595a1c3ec8e8aad6969aa should fix this. Shifted from GC_SCORE to IGC field to keep compatability with gtc2vcf workflow.

rajwanir commented:
> There are two identical files in the sub_workflows directory now.
> 
> * idat_intensity.smk
> * intensity_check.smk
> 
> ```
> sub_workflows$ ls
> contamination.smk  delivery.smk  entry_points.smk  idat_intensity.smk  intensity_check.smk  sample_qc.smk  subject_qc.smk
> ```

Removed in 9cda9ed5d48c0cea3656f3ec5ce1d29587c08da6

jaamarks commented:
> You should expect to see no difference except for multi-allellics not imported which is not supported by plink/1.9 and would need to be fixed upstream in preparation of bcf.
>
>Overall, I can open a separate issue and we can adress this more holistically throughout the pipeline in a different PR. However, for the BCF entry point, I see the transformation is as expected with no flip.

Thanks for the explanation @rajwanir. Can you open a separate issue for this explaining what you found?

rajwanir commented:
> > You should expect to see no difference except for multi-allellics not imported which is not supported by plink/1.9 and would need to be fixed upstream in preparation of bcf.
> > Overall, I can open a separate issue and we can adress this more holistically throughout the pipeline in a different PR. However, for the BCF entry point, I see the transformation is as expected with no flip.
> 
> Thanks for the explanation @rajwanir. Can you open a separate issue for this explaining what you found?

Opened #321 to reference this allele flip issue.

Commit:
Adds a config suffix validator for bcf_file.

jaamarks commented on src/cgr_gwas_qc/workflow/scripts/vcf2abf.py:
```diff
@@ -0,0 +1,142 @@
+                rec.ref[0],
+                rec.alts[0],
+                is_biallelic_snp(rec),
+                rec.info["GC_SCORE"],
```
Does this line need to be edited too? I got an error again about the `GC_SCORE` field.
```
rule pull_b_allele_freq_from_1kg:
    input: /mnt/nfs/gigantor/ifs/DCEG/Home/marksja/dev-tests/samples.bcf, /DCEG/Projects/DataDelivery/GWAS_Data_Delivery/CONFLUENCE_FB28648_05172021/testing_500samples/manifest/ALL.wgs.shapeit2_integrated_v1a.GRCh38.20181129.sites.vcf.gz
    output: sample_level/GSAv3Confluence.AF.abf.txt
    jobid: 0                                                                                                                      resources: mem_mb=19787, disk_mb=19787, tmpdir=/tmp                                                                       
/home/marksja/miniconda3/envs/cgr-dev/bin/python /mnt/nfs/gigantor/ifs/DCEG/Home/marksja/dev-tests/024/.snakemake/scripts/tmph3hp_w1c.vcf2abf.py
[E::idx_find_and_load] Could not retrieve index file for '/mnt/nfs/gigantor/ifs/DCEG/Home/marksja/dev-tests/samples.bcf'
Traceback (most recent call last):
  File "/mnt/nfs/gigantor/ifs/DCEG/Home/marksja/dev-tests/024/.snakemake/scripts/tmph3hp_w1c.vcf2abf.py", line 144, in <module>
    main(**defaults)  # type: ignore
  File "/mnt/nfs/gigantor/ifs/DCEG/Home/marksja/dev-tests/024/.snakemake/scripts/tmph3hp_w1c.vcf2abf.py", line 77, in main
    rec.info["GC_SCORE"],
  File "pysam/libcbcf.pyx", line 2598, in pysam.libcbcf.VariantRecordInfo.__getitem__
KeyError: 'Unknown INFO field: GC_SCORE'
[Fri Sep 13 15:08:08 2024]                                                                                                    Error in rule pull_b_allele_freq_from_1kg:
    jobid: 0                                                                                                                      output: sample_level/GSAv3Confluence.AF.abf.txt

```

rajwanir replied:
Yes, for compatibility with gtc2vcf prepared bcf. 1ee63adf5d118aae96def3b0dba6c37614df3488 should fix this. Thanks.

jaamarks replied:
I'm still having issues with this rule.
```
rule pull_b_allele_freq_from_1kg:
    input: /mnt/nfs/gigantor/ifs/DCEG/Home/marksja/dev-tests/samples.bcf, /DCEG/Projects/DataDelivery/GWAS_Data_Delivery/CONFLUENCE_FB28648_05172021/testing_500samples/manifest/ALL.wgs.shapeit2_integrated_v1a.GRCh38.20181129.sites.vcf.gz
    output: sample_level/GSAv3Confluence.AF.abf.txt
    jobid: 0
    resources: mem_mb=19787, disk_mb=19787, tmpdir=/tmp

/home/marksja/miniconda3/envs/cgr-dev/bin/python /mnt/nfs/gigantor/ifs/DCEG/Home/marksja/dev-tests/024/.snakemake/scripts/tmp3gubjs59.vcf2abf.py
[E::idx_find_and_load] Could not retrieve index file for '/mnt/nfs/gigantor/ifs/DCEG/Home/marksja/dev-tests/samples.bcf'
Traceback (most recent call last):
  File "/mnt/nfs/gigantor/ifs/DCEG/Home/marksja/dev-tests/024/.snakemake/scripts/tmp3gubjs59.vcf2abf.py", line 140, in <module>
    main(**defaults)  # type: ignore
  File "/mnt/nfs/gigantor/ifs/DCEG/Home/marksja/dev-tests/024/.snakemake/scripts/tmp3gubjs59.vcf2abf.py", line 73, in main
    rec.info["IGC"],
  File "pysam/libcbcf.pyx", line 2598, in pysam.libcbcf.VariantRecordInfo.__getitem__
KeyError: 'Unknown INFO field: IGC'
[Fri Sep 13 18:05:27 2024]
Error in rule pull_b_allele_freq_from_1kg:
    jobid: 0
    output: sample_level/GSAv3Confluence.AF.abf.txt

```

Here is the a view of part of the head of my `samples.bcf` input file. Let me know if there is anything else I can provide to help troubleshoot.

```
##bcftools_viewCommand=view samples.bcf; Date=Fri Sep 13 18:13:22 2024
#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT SC676000_PC45624_G12    SC676001_PC45625_D03 ...
```

rajwanir replied:
@jaamarks Thanks again for reviewing the PR. 
I just wanted to update that if you need to release a new version for confluence project you may go ahead without this PR should the need be. It may take some time to update the docs and write the unit tests. The jupyter at CCAD is also down.
The alternative would be if we could have this PR merged to the new version and I can update the docs and tests later. It might save some time realigning the code with the new version and we have already tested it on the real 500 samples dataset. However, I do see that having the docs and tests updated would be ideal. 
Thanks.

jaamarks replied:
Thanks for the update, @rajwanir. We believe it would be best to hold off on merging this PR until all comments have been addressed, including updates to the documentation and tests. If you have any questions, please feel free to reach out.
 
We also have another question: How much utility do you think adding a VCF entry point would provide? Do you think it is worth the effort to implement this feature and do you see it as a relatively straightforward addition once the BCF entry point is completed?

rajwanir replied:
@jaamarks Could you please try after this addea7083a996997304920a6c3dabb3c5d7326a8? I have exlcuded using the score field using this incompatibility issue in different ways of preparing bcf. I have tested on my end and it works fine. Let me know if you run into issue. Thanks.

rajwanir replied:
@jaamarks 

**For unit testing**

1. I have added three new unit testing scripts scripts in 077f3b1ea5db611e76d87df0be209f8725a939d2 which should test: `src/cgr_gwas_qc/workflow/scripts/vcf2abf.py` , `src/cgr_gwas_qc/workflow/scripts/median_intensity_from_vcf.py ` and `src/cgr_gwas_qc/workflow/scripts/vcf2adpc.py` .
2. I don't see any exisitng testing scripts that checks for any of the `grouped*` script so I have left `src/cgr_gwas_qc/workflow/scripts/grouped_intensity_from_vcf.py` and `src/cgr_gwas_qc/workflow/scripts/grouped_contamination.py`. I believe the rationale might be that these are essentially wrappers to execute main sripts in the parallel mode. I do see that those may be indirectly tested through the subworkflows.
3.  Much of the subworkflows is based on the `RealData`. I have synced the RealData from CCAD2. However, currently even with the default branch the pytest fails on my end. First, I note that  `_reference_fasta ` should be set in RealData since the gtc2vcf workflow is created. Second, I ran into another issue that conda envinorments seem inccorectly setup in this context. I am happy to work through additional subworkflow additions/modifications needed relevent to this PR after the default is fixed.
4. I have updated the FakeData repository with files needed to test bcf entry in a separate PR for the test repo here: https://github.com/NCI-CGR/GwasQcPipeline-test-data/pull/1. Please take a look and merge this first so you may be able to take a look at the new unit testing scripts added in this PR.

Please take a look and let me know if you have any questions. Thank you.

rajwanir replied:
@jaamarks 

I have updated the documentation to reflect the BCF entry point:

1. 1afc4e79ebd73e90d01a10ec949800853d9c7783 Indicates the bcf entry point in the user config 
2. cbd7c898fe16a542d61261d8890a359fa8b1ef7a Adds information about bcf entry point on the entry point page. I have added as point in the description since the entry point processing is essentially that it will convert to plink bed (similar to aggregated BED input)
3. 20b6d3074fe7a5fcddb4e1ac46e8ce2c93d2569b Updates the contamination subworkflow documentation to show now it is an isolated subworkflow and supports both GTC and BCF input.
4. b1f1c90ecf8a8c4f7cdaa75cad5594e68b3567c0 and 7236b041d872962592a93f78fe2ffdb536ed155f  Adds the information for intensity subworkflow and show processing for both GTC and BCF inputs.

Please let me know if there is anything else that would suggest to update/modify regarding this PR. Thank you.

jaamarks commented on src/cgr_gwas_qc/workflow/scripts/vcf2abf.py:
```diff
@@ -0,0 +1,142 @@
+    if abf_file is None:
+        abf_file = bcf_file.with_suffix(".abf.txt")
+
+    # vcf_file='/scratch/rajwanir2/data_catalog/QC-test/results/GSAMD-24v1-0-contamseries/vcf/GSAMD-24v1-0-contamseries.bcf'
```
Remove these debugging lines 57–59.

rajwanir replied:
Removed in 874eb15e1e0fac826a01450e7ed8898c63cc112d. Thanks.

jaamarks commented on src/cgr_gwas_qc/workflow/scripts/grouped_contamination.py:
```diff
@@ -98,12 +138,13 @@ def estimate_contamination(
-        outfile=snakemake.output[0],  # type: ignore # noqa
-        threads=snakemake.threads or 2,  # type: ignore # noqa
+        defaults = {}
+        defaults.update({k: Path(v) for k, v in snakemake.input.items() if not k.startswith("_")})  # type: ignore # noqa
```
Looks like Black is failing on this line. Can you address this?
<img width="1315" alt="image" src="https://github.com/user-attachments/assets/f39827c4-4166-47a3-9a1d-8a0ecdaa7806">

rajwanir replied:
2b8acbcd6056ded1b2e2a135051761bc094a0dd6 should fix this. Thanks.

jaamarks reviewed (changes requested):
# Tests
Could you please update the functional and unit tests for each script that you modified or created? This PR introduces four new scripts:
* `src/cgr_gwas_qc/workflow/scripts/grouped_intensity_from_vcf.py`
* `src/cgr_gwas_qc/workflow/scripts/median_intensity_from_vcf.py`
* `src/cgr_gwas_qc/workflow/scripts/vcf2abf.py`
* `src/cgr_gwas_qc/workflow/scripts/vcf2adpc.py`

Additionally, it modifies the existing script:
* `src/cgr_gwas_qc/workflow/scripts/grouped_contamination.py`

Since this PR also introduces a new sub-workflow, `intensity_check.smk`, by separating the intensity check from the contamination workflow, please ensure that any necessary updates are made to the testing file located at:
* `src/cgr_gwas_qc/workflow/sub_workflows/contamination.smk`

Furthermore, a new test for the `intensity_check.smk` sub-workflow should be created. You can use the existing test files in GwasQcPipeline/tests/workflow/sub_workflow as templates for this purpose.

And please add any other test you think would be helpful. 

<br><br>

# Docs
Could you please update the documentation with these new changes? You will probably be able to identify all the necessary places, but for example I see the following pages that need updated:
1. Configuration Section: https://nci-cgr.github.io/GwasQcPipeline/getting_started/configuration.html?highlight=vcf#user-files
2. Entry-points: https://nci-cgr.github.io/GwasQcPipeline/sub_workflows/entry_points.html
    * Additionally, please add a new figure illustrating the option for the BCF entry point.
3. Contamination Subworkflow: https://nci-cgr.github.io/GwasQcPipeline/sub_workflows/contamination.html
4. Intensity Check Subworkflow
    * add a new page for the `intensity_check.smk` subworkflow.

If there are any other areas that require updates, please address those as well. Thank you!

Commit:
Removes some debugging lines from vcf2abf.py

Commit:
Shifts to using IGC as genotype score for preparing abf using vcf2abf.
Consistent with vcf2adpc.py. Now both should IGC and work with vcf/bcf prepared with gtc2vcf workflow.

jaamarks commented on src/cgr_gwas_qc/workflow/scripts/vcf2adpc.py:
```diff
@@ -0,0 +1,143 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# In[1]:
```
remove commented line 4

rajwanir replied:
Removed in ffdcf3fe68e81fc6e28fd295d969c81a0cc28fdd.

jaamarks commented on src/cgr_gwas_qc/workflow/scripts/vcf2adpc.py:
```diff
@@ -0,0 +1,143 @@
+# from more_itertools import unzip
+from cgr_gwas_qc.parsers.illumina import AdpcRecord, AdpcWriter
+
+# In[ ]:
```
remove commented line 19

rajwanir replied:
Removed in ffdcf3fe68e81fc6e28fd295d969c81a0cc28fdd

jaamarks commented on src/cgr_gwas_qc/workflow/scripts/vcf2adpc.py:
```diff
@@ -0,0 +1,143 @@
+            fh.write(record)
+
+
+# In[2]:
```
remove commented line 102

rajwanir replied:
removed in 5c6032cf3f86220df131476abeade127631b8278

jaamarks commented on src/cgr_gwas_qc/workflow/scripts/vcf2adpc.py:
```diff
@@ -0,0 +1,143 @@
+        yield AdpcRecord(x_raw, y_raw, x_norm, y_norm, genotype_score, genotype)
+
+
+# In[ ]:
```
remove commented line 118, along with any extra empty lines.

rajwanir replied:
removed in 5c6032cf3f86220df131476abeade127631b8278

jaamarks commented on src/cgr_gwas_qc/workflow/scripts/vcf2adpc.py:
```diff
@@ -0,0 +1,143 @@
+#     return(pd.DataFrame(adpc_records,columns=['x_raw', 'y_raw', 'x_norm', 'y_norm', 'genotype_score', 'genotype']))
+
+
+# In[ ]:
```
Remove these commented lines.

rajwanir replied:
removed in 5c6032cf3f86220df131476abeade127631b8278

Commit:
Removes commented lines from vcf2adpc.py and vcf2abf.py

Commit:
fixes a black syntax issue in importing snakemake params in grouped_contamatination.py

Commit:
Modifies vcf2abf to not use gentrain score in preparing abf.
Earliar gentrain score was used to mark AF as NA if score is negative. This change excludes using it to ensure compatibility with gtc2vcf workflow prepared bcf.

Commit:
Includes a None case in the BCF config validator to work with unit testing.

Commit:
Adds the app mode in median_intensity_from_vcf to work directly as stand-alone.

Commit:
Adds the new unit test scripts relevent to BCF entry

Commit:
docs: updates docstring to show bcf as a user_file in config

Commit:
docs: Updates the entry_point documentation to show BCF as an entry_point.

Commit:
docs: Updates contamination subworkflow documentation to show processing with BCF input.

Commit:
docs: Adds the documentation for new intensity subworkflow and shows processing for both IDAT and BCF input.

Commit:
docs: Updates index to include intensity_check subworkflow.

jaamarks commented on docs/sub_workflows/intensity_check.rst:
```diff
@@ -0,0 +1,27 @@
+   :width: 45%
+
+The Intenstiy check sub-workflow.
+This workflow will caclucate the median intensity for each sample individually.
```
spell check

rajwanir replied:
12f73c8f80c84c574ece8a335434b94a5e0de919 should resolve. Thanks.

jaamarks commented on src/cgr_gwas_qc/workflow/scripts/grouped_intensity_from_vcf.py:
Remove Jupyter-like comments from this file. Example:
```
# In[5]:

...

# In[10]:
```

rajwanir replied:
removed in 5c6032cf3f86220df131476abeade127631b8278

jaamarks commented on docs/sub_workflows/contamination.rst:
```diff
@@ -6,27 +6,28 @@ Contamination Sub-workflow
+This workflow will estimate contamination using verifyIDintensity on each sample individually.
+It requires that you have aggregated BCF or GTC files.
+It first pulls B-allele frequencies from the 1000 Genomes VCF file.
+It then estimate contamination for each sample and aggregates these results.
```
should be: "It then estimates contamination..."

rajwanir replied:
fixed in 12f73c8f80c84c574ece8a335434b94a5e0de919. Thanks.

jaamarks commented on tests/workflow/scripts/test_median_intensity_from_vcf.py:
```diff
@@ -0,0 +1,27 @@
+
+"""Test calculation of median intensity from BCF
+
+We need to assess whether the intensity meets the miniumum threshold. For this purpose, we typically check median IDAT intensity. Here we attempt to test the same with BCF as input file.
```
spell check: minimum

rajwanir replied:
fixed in 12f73c8f80c84c574ece8a335434b94a5e0de919. Thanks.

jaamarks reviewed (changes requested):
Hi @rajwanir  - looks like were getting close. There were some minor issues I am requesting to fix. 

I merged the PR for GwasQcPipeline-test-data for the FakeData. And updated this submodule for the superproject. All relevant tests passed after this.

Commit:
docs: corrects spellings in documentation

Commit:
Removes jupyter comment lines from scripts.

Commit:
docs: updates the docstring for user_files in config section to include BCF.

Commit:
docs: redo updates the user_files docstring to include BCF.

jaamarks reviewed (approved):
LGTM
